### ISSUE TITLE
object location and detection

### ISSUE THREAD

**Issue by kaishijeng** [dmlc/mxnet#214]
HI,

  fast-rcnn or fcatser-rcnn is a very useful application. Both are implemented and  trained in Caffe framework.
 Is there a way to import its network model and pretrained model to mxnet? If not. do you have a plan to use mxnet to  implement fast-rcnn or faster-rcnn?

Thanks,

Kaishi

**Comment by sunshineatnoon** [comment]
Same  thought here.

**Comment by antinucleon** [comment]
I will add feature to write handcraft kernel directly in op,  and we always welcome you to send PR to contribute to MXNet

**Comment by sunshineatnoon** [comment]
@antinucleon Thanks!

**Comment by futurely** [comment]
Fast and Faster R-CNN's changes to Caffe are all in [this commit](
https://github.com/rbgirshick/caffe-fast-rcnn/commit/4115385deb3b907fcd428ac0ab53b694d741a3c4) and the object detection application is [here](https://github.com/rbgirshick/py-faster-rcnn).
```
Fast and Faster R-CNN change set
 - smooth l1 loss
 - roi pooling
 - expose phase in pycaffe
 - dropout scaling at test time (needed for MSRA-trained ZF network)
```

**Comment by kaishijeng** [comment]
Any chance to add layers which are required by faster-rcnn in near future?

**Comment by loweew** [comment]
are there any plans for a faster-rcnn type example in the image-classification section? This would be greatly appreciated, if possible.

**Comment by precedenceguo** [comment]
New operator `mxnet.symbol.ROIPooling` is slightly tricky. Here are some key information that are missing from documentation.
```python
import mxnet as mx
data = mx.sym.Variable('data')
# [batch_size, channel, height, width]
rois = mx.sym.Variable('rois')
# [batch_size, roi_number, 5]
# last dimension is [batch index of image, x1, y1, x2, y2]
# some convolutional layer
roi_pool = mx.sym.ROIPooling(data=data, rois=rois, pooled_size=(6, 6), spatial_scale=0.0625)
# please note that batch_size changes from batch_size to batch_size * roi_number after ROI pooling.
```
It should be possible to drop the `batch_index` from the `rois` data since images and rois are aligned to each batch. But currently this remains unsolved.

**Comment by jonbakerfish** [comment]
@precedenceguo  How to train network with `mxnet.symbol.ROIPooling`? Any example?

**Comment by jonbakerfish** [comment]
I noticed that the [executor_manager.DataParallelExecutorGroup](https://github.com/dmlc/mxnet/blob/master/python/mxnet/executor_manager.py#L230) uses the same `slices` for both input data and labels. But in the case of fast-rcnn, during training, the input images are in shape (2,3,H,W) while the labels' are `(128,) for the 128 ROIS `. How can we change the code for training?

**Comment by precedenceguo** [comment]
Multiple devices training split data into slices for devices. In this example, each data batch has image shape `(2, 3, H, W)` and label shape `(128, )`. Splitting data batches does not compromise each batch since loader supply data and label together in each batch. Therefore no change is necessary except the number of training devices. For the training of Fast R-CNN, it is recommended to see the [original caffe version](https://github.com/rbgirshick/caffe-fast-rcnn).

**Comment by jonbakerfish** [comment]
I'm not using multiple devices for training. I use the python api [`model.fit`](https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L694) which actually calls `_train_multi_device`. The [executor_manager](https://github.com/dmlc/mxnet/blob/master/python/mxnet/executor_manager.py#L230) inside `_train_multi_device` uses the same `slices` for both input data and labels. It works for the case where the batch sizes are the same for both input data and label. I'm looking for advice on how to change the code accordingly when input data and label have different batch sizes.

**Comment by precedenceguo** [comment]
There could exist some issue about varying batch size with the FeedForward API. Is there any error message or anomaly?

**Comment by tqchen** [comment]
c.f. https://github.com/dmlc/mxnet/tree/master/example/rcnn
